src/basics.ml

```diff
@@ -201,6 +208,103 @@ let has_metas : term -> bool =
+  let mk = fresh_meta mk_typ k in
+  let tk = _Meta mk (Array.map _Vari vs) in
+  Bindlib.unbox (build_prod k tk)
+
+(** [new_symb name t] returns a new function symbol of name [name] and of
+    type [t]. *)
+let new_symb name t =
+  { sym_name = name ; sym_type = ref t ; sym_path = [] ; sym_def = ref None
+  ; sym_impl = [] ; sym_rules = ref [] ; sym_mode = Const }
+
+(** [to_m k metas t] computes a new (boxed) term by replacing every pattern
+    variable in [t] by a fresh metavariable and store the latter in [metas],
+    where [k] indicates the order of the term obtained *)
+let rec to_m : int -> (meta option) array -> term -> tbox = fun k metas t ->
+  match unfold t with
+  | Vari x         -> _Vari x
+  | Symb (s, h)    -> _Symb s h
+  | Abst (a, t)    ->
+      let (x, t) = Bindlib.unbind t in
+      _Abst (to_m 0 metas a) (Bindlib.bind_var x (to_m 0 metas t))
+  | Appl (t, u)    -> _Appl (to_m (k + 1) metas t) (to_m 0 metas u)
+  | Patt (i, n, a) ->
+      begin
+        let a = Array.map (to_m 0 metas) a in
+        let l = Array.length a in
+        match i with
+        | None   ->
+            let m = fresh_meta ~name:n (build_meta_type (l + k)) l in
+            _Meta m a
+        | Some i ->
+            match metas.(i) with
+            | Some m -> _Meta m a
+            | None   ->
+                let m = fresh_meta ~name:n (build_meta_type (l + k)) l in
+                metas.(i) <- Some m;
+                _Meta m a
+      end
+  | _              -> assert false
+
+exception Not_FO
+
+(** [to_closed symbs t] computes a new (boxed) term by replacing every
+    pattern variable in [t] by a fresh symbol [c_n] of type [t_n] ([t_n] is
+    another fresh symbol of type [Kind]) and store [c_n] the latter in
+    [symbs]. *)
+let rec to_closed : (sym option) array -> term -> tbox
+  = fun symbs t ->
+  match unfold t with
+  | Vari x            -> _Vari x
+  | Symb (s, h)       -> _Symb s h
+  | Abst (a, t)       ->
+      let (x, t) = Bindlib.unbind t in
+      _Abst (to_closed symbs a) (Bindlib.bind_var x (to_closed symbs t))
+  | Appl (t, u)       -> _Appl (to_closed symbs t) (to_closed symbs u)
+  | Patt (i, n, [||]) ->
+      begin
+        match i with
+        | None   ->
+            let t_n = new_symb ("{t_" ^ n) Kind in
+            let term_t_n = symb t_n in
+            let c_n = new_symb ("{c_" ^ n) term_t_n in
+            _Symb c_n Nothing
+        | Some i ->
+            match symbs.(i) with
+            | Some s -> _Symb s Nothing
+            | None   ->
+                let t_n = new_symb ("{t_" ^ n) Kind in
+                let term_t_n = symb t_n in
+                let c_n = new_symb ("{c_" ^ n) term_t_n in
+                symbs.(i) <- Some c_n;
+                _Symb c_n Nothing
+      end
+  | Patt _            -> raise Not_FO
+  | _                 -> assert false
+
+(** [is_new_symb s] checks if [s] is a function symbol created for checking
+    SR. *)
+let is_new_symb s = s.sym_name.[0] = '{'
```

Comment:
This is too adhoc. When you replace pattern variables par symbols, you get an array with the new symbols. Use this array instead.

Reply:
Knowing that there are lots of intermediate functions written for the completion procedure, I am not sure if it is a good idea to use the array of symbols. Each of these functions should take the array as an argument in order to to this. Another idea is to write a local function for it instead of including it in basics.ml as it is only used in three functions.

Reply:
But I don't understand why you need to distinguish between old and new symbols in the first place... You need to clarify this first.